src/threads/table.ts

```diff
@@ -0,0 +1,66 @@
+import { DynamoDB } from '@aws-sdk/client-dynamodb';
+import * as D from 'io-ts/Decoder';
+import { fromDBItem, toDBItem } from '../dynamodb';
+import { decode } from '../decoder';
+
+const baseEventDecoder = D.struct({
+  threadId: D.string,
+  timestamp: D.number,
+});
+
+const threadOpenedEventDecoder = D.struct({
+  type: D.literal('thread_opened'),
+  byUserId: D.string,
+});
+
+const threadClosedEventDecoder = D.struct({
+  type: D.literal('thread_closed'),
+  byUserId: D.string,
+});
+
+const threadEventInput = D.union(threadOpenedEventDecoder, threadClosedEventDecoder);
+type ThreadEventInput = D.TypeOf<typeof threadEventInput>;
+
+const threadEventDecoder = D.intersect(baseEventDecoder)(threadEventInput);
+export type ThreadEvent = D.TypeOf<typeof threadEventDecoder>;
```

Comment:
I don't get this. 
What you want is a `ThreadOpenedEvent | ThreadClosedEvent` so it is the union.
I don't get what ` D.intersect(baseEventDecoder)(threadEventInput)` corresponds to... 1) I am not sure it is the way intersect is [supposed to be used](https://github.com/gcanti/io-ts/blob/master/Decoder.md#the-intersect-combinator), 2) I don't understand why `threadEventInput` is not good here.

Reply:
`threadEventInput` does not include base properties `pk` and `time` (former `threadId` and `timestamp`)
I initially wrote interfaces `interface ThreadOpenedEvent extends BaseEvent {...}` but I found out that the `Omit<ThreadEvent, keyof BaseEvent>` is incorrect when I'll add other events like `{ type: 'follow', connectionId: string }`
So for the method `addThreadEvent()` I had to consider only input event without base properties because base properties are computed and should not to be provided.
And the intersect is the way to re-add the 2 base properties `pk` and `timestamp`.

If it's unclear we can also have a call

Reply:
Actually, I think I misread the types and that's why I had difficulties understanding what you were doing. Ok I got it and I understand why you need both types.
What's still a bit strange to me is that you bind `threadEventInput` to the event-specific attributes... but it is just a coincidence that the "common attributes"  (baseEventDecoder) are not in   `threadEventInput` ... if we add, let's say "validity" to all entries... it may be included in  `threadEventInput`.  I mean semantically,  in `addThreadEvent`, instead of `threadEventInput`, what you need is more `Omit<ThreadEvent,'pk'|'time'>` I think. Anyway, it may just be a change later.